### PR TITLE
COS: Fix tools replicate

### PR DESCRIPTION
**Pull request by hanjm**
Fix issue reported by @wu3396 (https://github.com/thanos-io/thanos/pull/5139#issuecomment-1040001559 ), pr #5139 introduce multi-part upload, it uses TryToGetSize to get the size of `reader`, but it will error when the reader is wrapped by TimingReadCloser. this pr continue to fix it.

1. Fix `TryToGetSize` error when the reader wrappered by `TimingReadCloser`, It will also fix `cos/s3/bos/swift`  get size error when use `objstore.TryToGetSize(r)`. 
2. Add size info into reader returned by `(*bucket).Get` for pass it to `(*bucket).Upload` function.

Signed-off-by: Jimmie Han <hanjinming@outlook.com>

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

1. Implement ObjectSizer interface for timingReadCloser.
2. Add unit test.
## Verification

1. unit test.
2. use `thanos tool replicate` to reproduce the issue and verify.

### CODE REVIEW

**Comment by hanjm** [comment]
docs check fail, it will fixed by #5167 (:

**Comment by yeya24** [comment]
Can you please rebase main? The doc should be fixed already on main.

**Review by yeya24 — DISMISSED** [review]
Looks like https://github.com/thanos-io/thanos/pull/4687/files is overlapped with this pr.

**Comment by hanjm** [comment]
> Details

Oh, yes.

**Comment by hanjm** [comment]
> Can you please rebase main? The doc should be fixed already on main.

Thank you reviewing. 🚀  Rebase Done.

**Review by yeya24 — APPROVED** [review]
LGTM.

**Comment by hanjm** [comment]
@yeya24 some e2e test is flaky, I retry it. Please review again, thank you.

File: pkg/objstore/cos/cos.go [hunk 1]
```diff
@@ -333,8 +333,19 @@ func (b *Bucket) getRange(ctx context.Context, name string, off, length int64) (
 		runutil.ExhaustCloseWithLogOnErr(b.logger, resp.Body, "cos get range obj close")
 		return nil, err
 	}
+	// Add size info into reader for pass it to Upload function.
```
Comment (matej-g) [inline pkg/objstore/cos/cos.go:336]:
```suggestion
	// Add size info into reader to pass it to Upload function.
```

**Review by matej-g — APPROVED** [review]
Looks good! :+1:

File: pkg/objstore/cos/cos.go (hunk 1 above)
Comment (hanjm) [reply pkg/objstore/cos/cos.go]: Done

**Review by GiedriusS — APPROVED** [review]

**Pull request merged**